### PR TITLE
jalmah.cpp : Cleanups

### PR DESCRIPTION
**Pull request by cam900**
jalmah.cpp : Cleanup ACCESSING_BITS, Cleanup duplicates, Reduce runtime tag lookups, Fix/Cleanup background tile bank in urashima, Add notes, Configure'd OKI banking, Add save state

### CODE REVIEW

File: src/mame/drivers/jalmah.cpp [hunk 1]
```diff
@@ -238,11 +248,11 @@ class urashima_state : public jalmah_state
 	required_shared_ptr_array<uint16_t, 2> m_videoram;
 	required_device<gfxdecode_device> m_gfxdecode;
 	
-	uint16_t m_vreg[2][0x800];
+	std::unique_ptr<uint16_t[]> m_vreg[2];
 	tilemap_t *m_layer[2];
 	
 	TILEMAP_MAPPER_MEMBER(range0_16x16);
-	TILEMAP_MAPPER_MEMBER(range3_8x8);
+	TILEMAP_MAPPER_MEMBER(range1_8x8);
```
Comment (angelosa) [inline src/mame/drivers/jalmah.cpp]: range3 is an intentional naming: it follows up what was this function name before tilemap rewrite, it's not referred to the layer name itself.

File: src/mame/drivers/jalmah.cpp [hunk 2]
```diff
@@ -253,18 +263,22 @@ Video Hardware start
...
 	int code = m_videoram[TileChip][tile_index];
 	uint16_t tile = code & 0xfff;
-	
-	if(TileChip == 0)
-		tile |= m_tile_bank << 12;
-	
+
+	// Tile bank is only affected when bit 11 is on,
+	// and Each bank is 256KB(0x40000) each
+	if ((TileChip == 0) && (tile & 0x800))
```
Comment (angelosa) [inline src/mame/drivers/jalmah.cpp]: This is a change in functionality, do you have proof about this?

File: src/mame/drivers/jalmah.cpp [hunk 3]
```diff
@@ -1174,19 +1207,15 @@ static const gfx_layout tilelayout =
...
+	{ STEP16(0,4*8) },
 	32*32
 };
 
 static GFXDECODE_START( gfx_urashima )
 	GFXDECODE_ENTRY( "scroll0", 0, tilelayout, 0x100, 16 )
-	GFXDECODE_ENTRY( "scroll3", 0, charlayout, 0x000, 16 )
-	GFXDECODE_ENTRY( "scroll1", 0, tilelayout, 0x100, 16 )
-	GFXDECODE_ENTRY( "scroll2", 0, tilelayout, 0x100, 16 )
+	GFXDECODE_ENTRY( "scroll1", 0, charlayout, 0x000, 16 )
```
Comment (angelosa) [inline src/mame/drivers/jalmah.cpp]: This was left intentionally. 
What's the MACHINE_UNEMULATED_PROTECTION part you don't understand?

File: src/mame/drivers/jalmah.cpp [hunk 4]
```diff
@@ -289,10 +303,16 @@ void urashima_state::video_start()
...
 	// also it's confirmed to be 64 x 64 with 2nd tier girls stripping
-	m_layer[1] = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(FUNC(urashima_state::get_tile_info_urashima<1>),this),tilemap_mapper_delegate(FUNC(urashima_state::range3_8x8),this),8,8,64,64);
+	m_layer[1] = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(FUNC(urashima_state::get_tile_info_urashima<1>),this),tilemap_mapper_delegate(FUNC(urashima_state::range1_8x8),this),8,8,64,64);
 
 	for(int i=0;i<2;i++)
 		m_layer[i]->set_transparent_pen(15);
+
+	for (int layer = 0; layer < 2; layer++)
+	{
+		m_vreg[layer] = make_unique_clear<uint16_t[]>(0x800);
```
Comment (angelosa) [inline src/mame/drivers/jalmah.cpp]: This is 0x400 not 0x800 (also can be safely allocated as shared pointer I guess)

**Comment by cam900** [comment]
In current driver, urashima Tilemap ROM loading have some duplicate; I guess only tilemap code 0x800-0xfff is bank switchable, and This PR has save states for variables.

File: src/mame/drivers/jalmah.cpp [hunk 5]
```diff
@@ -126,18 +126,22 @@ OSC:    12.000MHz
...
 		m_jm_shared_ram(*this, "jshared_ram"),
 		m_jm_mcu_code(*this, "jmcu_code"),
-		m_maincpu(*this, "maincpu")
-		{ }
+		m_maincpu(*this, "maincpu"),
+		m_prirom(*this, "prirom"),
+		m_okibank(*this, "okibank"),
+		m_key_io(*this, "KEY%u", 0U),
+		m_system_io(*this, "SYSTEM"),
+		m_dsw_io(*this, "DSW") { }
```
Comment (cuavas) [inline src/mame/drivers/jalmah.cpp]: Please don't move the braces to the end of the line when the initialiser list isn't all on one line like this.

**Comment by angelosa** [comment]
Yeah, and it's called cluttering in tilemap code. 
I prefer having those loaded in rom loading and repeated than having a conditional branch inside the drawing itself (-> every tile changed will do an extra check).
Or, even better, use the rom region numbering instead (which **I will** eventually do but have more pressing stuff to do first).

**Comment by angelosa** [comment]
More or less applied with 89da3a3fdbed90d8d64ba0144d9e35fff65cef0b

Closing this cause being outdated.

**Pull request closed without merging**